File: javascript/0238-product-of-array-except-self.js
/**
 * Array
 * Time O(N) | Space O(N)
 * https://leetcode.com/problems/product-of-array-except-self/
 * @param {number[]} nums
 * @return {number[]}
 */
 function productExceptSelf(nums) {
    const result = [];
    let prefix = 1;
    let postfix = 1;
    
    for (let i = 0; i < nums.length; i++) {
        result[i] = prefix;
        prefix *= nums[i];
    }
    for (let i = nums.length - 2; i >= 0; i--) {
        postfix *= nums[i + 1];
        result[i] *= postfix;
    }
    
    return result;
};

File: scala/0238-product-of-array-except-self.scala
object Solution {
    def productExceptSelf(nums: Array[Int]): Array[Int] = {

        val result = Array.fill[Int](nums.length)(1)

        var prefix = 1
        for(i <- nums.indices) {
            result(i) = prefix
            prefix = prefix * nums(i)
        }

        var postfix = 1
        for(i <- nums.length-1 to 0 by -1) {
            result(i) = result(i) * postfix
            postfix = postfix * nums(i)
        }

        result
    }
}